File: src/main/java/pcb/uwu/impl/fundamental/Meters.java
package pcb.uwu.impl.fundamental;

import pcb.uwu.amounts.fundamental.Length;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.fundamental.LengthUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.units.fundamental.LengthUnit.METER;

public class Meters extends Length {

	// region constructors

	public Meters(Number value) {
		super(value, METER);
	}

	public Meters(Number value, Magnitude magnitude) {
		super(value, magnitude, METER);
	}

	public Meters(String value) {
		super(value, METER);
	}

	public Meters(String value, Magnitude magnitude) {
		super(value, magnitude, METER);
	}

	public Meters(BigDecimal value) {
		super(value, METER);
	}

	public Meters(BigDecimal value, Magnitude magnitude) {
		super(value, magnitude, METER);
	}

	public Meters(BigDecimalAmount amount) {
		super(amount, METER);
	}

	public Meters(BigDecimalAmount amount, Magnitude magnitude) {
		super(amount, magnitude, METER);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Meters plus(UnitAmount<LengthUnit> other, MathContext mathContext) {
		return new Meters(plusAmount(other, mathContext));
	}

	@Override
	public Meters minus(UnitAmount<LengthUnit> other, MathContext mathContext) {
		return new Meters(minusAmount(other, mathContext));
	}

	@Override
	public Meters multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Meters(multipliedByScalar(other, mathContext));
	}

	@Override
	public Meters dividedBy(BigDecimal other, MathContext mathContext) {
		return new Meters(dividedByScalar(other, mathContext));
	}

	// endregion
}

File: implementation/src/main/java/pcb/uwu/amount/base/Meters.kt
package pcb.uwu.amount.base

import pcb.uwu.amount.derived.geometry.area.SquareMeters
import pcb.uwu.amount.derived.mechanics.Force
import pcb.uwu.amount.derived.thermodynamics.Joules
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.LengthUnit
import pcb.uwu.unit.base.MeterUnit
import pcb.uwu.unit.derived.mechanics.NewtonUnit

class Meters : Length
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, MeterUnit)

    @JvmOverloads
    constructor(value: String,
                magnitude: Magnitude = NATURAL)
            : super(value, magnitude, MeterUnit)

    // region UnitAmount

    override fun plus(length: UnitAmount<LengthUnit>) =
        Meters(this.amount + (length to MeterUnit).amount)

    override fun minus(length: UnitAmount<LengthUnit>) =
        Meters(this.amount - (length to MeterUnit).amount)

    override fun times(scalar: Number) =
        Meters(this.amount * scalar)

    override fun div(scalar: Number) =
        Meters(this.amount / scalar)

    // endregion

    // region composition

    override fun times(length: Length) =
        SquareMeters(this.amount * (length to MeterUnit).amount)

    override fun times(force: Force) =
        Joules(this.amount * (force to NewtonUnit).amount)

    // endregion
}
